src/map.mo

```diff
@@ -0,0 +1,290 @@
+      let { leaves; nodes } = base.regions();
+
+      let ?(added, leaf) = base.put_(nodes, leaves, key) else return null;
+      let ret_value = if (added) {
+        base.setValue(leaves, leaf, value);
+        value;
+      } else {
+        let old_value = base.getValue(leaves, leaf);
+        base.setValue(leaves, leaf, value);
+        old_value;
+      };
+      ?(ret_value, Nat64.toNat(leaf));
+    };
+
+    /// Add `key` and `value` to enumeration.
+    /// Returns null if pointer size limit exceeded.
+    /// Lookup value if key is already present. First return value `size` is if the key in new to the enumeration
+    /// or index of key in enumeration otherwise. Second return is old value if new wasn't added or a new one otherwise.
+    ///
+    /// Example:
+    /// ```motoko
+    /// let e = StableTrie.Map({
+    ///   pointer_size = 2;
+    ///   aridity = 2;
+    ///   root_aridity = null;
+    ///   key_size = 2;
+    ///   value_size = 1;
+    /// });
+    /// assert(e.lookupOrPut("abc", "a") == ?("a", 0);
+    /// assert(e.lookupOrPut("aaa", "b") == ?("b", 1));
+    /// assert(e.lookupOrPut("abc", "c") == ?("a", 0);
+    /// ```
+    /// Runtime: O(key_size) acesses to stable memory.
+    public func lookupOrPut(key : Blob, value : Blob) : ?(Blob, Nat) {
+      let { leaves; nodes } = base.regions();
+
+      let ?(added, leaf) = base.put_(nodes, leaves, key) else return null;
+      let ret_value = if (added) {
+        base.setValue(leaves, leaf, value);
+        value;
+      } else {
+        base.getValue(leaves, leaf);
+      };
+      ?(ret_value, Nat64.toNat(leaf));
+    };
+
+    /// Returns `?(index, value)` where `index` is the index of `key` in order it was added to enumeration and `value` is corresponding value to the `key`,
+    /// or `null` it `key` wasn't added.
+    ///
+    /// Example:
+    /// ```motoko
+    /// let e = StableTrie.Map({
+    ///   pointer_size = 2;
+    ///   aridity = 2;
+    ///   root_aridity = null;
+    ///   key_size = 2;
+    ///   value_size = 1;
+    /// });
+    /// assert(e.put("abc", "a") == ?0);
+    /// assert(e.put("aaa", "b") == ?1);
+    /// assert(e.lookup("abc") == ?("a", 0);
+    /// assert(e.lookup("aaa") == ?("b", 1));
+    /// assert(e.lookup("bbb") == null);
+    /// ```
+    /// Runtime: O(key_size) acesses to stable memory.
+    public func lookup(key : Blob) : ?Blob {
+      Option.map<(Blob, Nat), Blob>(base.lookup(key), func(a) = a.0);
+    };
+
+    public func delete(key : Blob) : ?Blob {
+      let { leaves; nodes } = base.regions();
+      let bytes = Blob.toArray(key);
+
+      let idx = base.keyToRootIndex(bytes);
+      let child = base.getChild(nodes, 0, idx);
+      let (value, leaf) = deleteRec(nodes, leaves, key, bytes, child, base.root_bitlength);
+      if (leaf != child) {
+        base.setChild(nodes, 0, idx, leaf);
+      };
+      value;
+    };
+
+    func singleLeaf(region : Base.Region, node : Nat64) : Nat64 {
```

Comment:
`singleLeaf` is always called with `region = nodes`. Can we just hard-code `nodes` in it and remove the `region` argument?